{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7270bd6e-0bec-48f1-907b-af1f25874669",
   "metadata": {},
   "source": [
    "\n",
    "### map csrAs in motility experiment to species in human microbiome  \n",
    "Two methods:\n",
    "1) via Taxonomy  \n",
    "    -get taxonomic IDS from both the csrAs in the motility experiments and get taxonomic IDS from all of the species in the human gut microbiome\n",
    "2) via sequence similarity\n",
    "    - find all csrA orthologs in the Hs gut bacterial and archeal species and find if there are any exact matches or close matches\n",
    "1) get motility csrAs\n",
    "2) get taxonomy info from "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "954474dc-b719-498c-af82-d2e4e9863e3a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import Phylo\n",
    "import pandas as pd\n",
    "from jw_utils import jw_draw_tree as jdt\n",
    "from jw_utils import parse_fasta as pfa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "896497de-e2ad-4ec8-919d-5a52bb50abc3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aa12ea2e-6147-44f5-9ebc-090c36cc46f2",
   "metadata": {
    "tags": []
   },
   "source": [
    "### get metedata for human gut microbiome  \n",
    "FTP downloads:  \n",
    "http://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-gut/v2.0.1/  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b94c2d94-8af4-405a-8365-14d12ecf9c21",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "metadata_df =  pd.read_csv('../gut_microbiome/genomes-all_metadata.tsv', sep = '\\t')\n",
    "metadata_df = metadata_df.set_index('Genome')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5618348c-e3de-4671-980d-5eb04edf98ce",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Get bacterial and archeal phylogenetic trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8f1fc82c-b406-40fd-bc11-d5047a76e198",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#paths\n",
    "tree_dir = '../Hs_gut_phylotrees/'\n",
    "path_to_bac_tree = os.path.join(tree_dir, 'bac120_iqtree.nwk')\n",
    "path_to_arch_tree = os.path.join(tree_dir, 'ar122_iqtree.nwk')\n",
    "path_to_csra_tsv = '../Data/csrs.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f806895d-8c38-40ae-952d-954d3145dee9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: ftp.ebi.ac.uk\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: ftp.ebi.ac.uk\n"
     ]
    }
   ],
   "source": [
    "#get phylogenetic tree for gut microbiome\n",
    "if not os.path.exists(tree_dir):\n",
    "    os.makedirs(tree_dir)\n",
    "\n",
    "!curl http://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-gut/v2.0.1/phylogenies/bac120_iqtree.nwk > $path_to_bac_tree\n",
    "!curl http://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-gut/v2.0.1/phylogenies/ar122_iqtree.nwk > $path_to_arch_tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e14448e-c28a-4273-ab29-2d179fdba14b",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Get dictionaries that contian best hits for both the forward and reverse blasts for both subtilis and aeruginosa csrA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9c61a449-b675-4efd-9051-30182058797f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "def best_hit_to_dict(path_to_rblast_results, best_hit_col_name=None):\n",
    "    \"\"\"Return a dict with the accession of query genome and the best hit in the source genome\n",
    "\n",
    "    These are the results of the reciprical blast. For each genome, there was a best hit in the initial\n",
    "    blastp. This best hit was reciprical blasted against the genome containing the original query. If best\n",
    "    hit in this reciprical blastp is the initial query, then these proteins are likely orthologs.\n",
    "    \n",
    "    return (df): cols = [accession, protein_id]\n",
    "    \"\"\"\n",
    "    files = os.listdir(path_to_rblast_results)\n",
    "    paths = [os.path.join(path_to_rblast_results, file) for file in files]\n",
    "    d = {}\n",
    "    for path, file  in zip(paths, files):\n",
    "        accession = file.replace('_blastpOut.txt', '')\n",
    "        with open(path, 'r') as f:\n",
    "            for i,line in enumerate(f):\n",
    "                if line.startswith('>'):\n",
    "                    prot_id = line.split(' ')[0][1:]\n",
    "                    d[accession]=prot_id\n",
    "                    break\n",
    "    df = pd.DataFrame.from_dict(d, orient= 'index')\n",
    "    if not best_hit_col_name:\n",
    "        best_hit_col_name = 'besthit_prot_id'\n",
    "    df.columns = [best_hit_col_name]               \n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "best_rHits_subt_df = best_hit_to_dict('../all_rblast_results_subtilis/reciprocal_blast_results/',best_hit_col_name='rbest_hit_subt' )\n",
    "best_fHits_subt_df = best_hit_to_dict('../all_rblast_results_subtilis/forward_blast_results/', best_hit_col_name='fbest_hit_subt')\n",
    "best_rHits_pseudo_df = best_hit_to_dict('../all_rblast_results_Psuedo/reciprocal_blast_results/', best_hit_col_name='rbest_hit_pseudo')\n",
    "best_fHits_pseudo_df = best_hit_to_dict('../all_rblast_results_Psuedo/forward_blast_results/', best_hit_col_name='fbest_hit_pseudo')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4712ab9c-fffa-4f46-b25e-ec95d09fe004",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Get two dfs of hs gut associeated species accessoins  \n",
    "1) containing those species with doubly-verified csrAs (csrA ortholog in both subtilis AND aeuruginosa reciprocal best hit blast)\n",
    "2) second list containing all species who had ortholog in at least one (subtilis OR aeuruginosa eciprocal best hit blast)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "df65fd7c-01fc-4497-8b5e-76b0f89ec68e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rbest_hit_subt</th>\n",
       "      <th>fbest_hit_subt</th>\n",
       "      <th>rbest_hit_pseudo</th>\n",
       "      <th>fbest_hit_pseudo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MGYG000003619</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000003619_01053</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000003619_01053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000003060</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000003060_00284</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000003060_00284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000000626</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000000626_00410</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000000626_00410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002462</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000002462_03497</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000002462_03497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002532</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000002532_03127</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000002532_03127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000000864</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000000864_02155</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000000864_02155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000001491</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000001491_01283</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000001491_01283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000001312</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000001312_01817</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000001312_01817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000001486</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000001486_01494</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000001486_01494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002404</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000002404_03669</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000002404_03669</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>599 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              rbest_hit_subt       fbest_hit_subt rbest_hit_pseudo  \\\n",
       "MGYG000003619    NP_391417.1  MGYG000003619_01053      NP_249596.1   \n",
       "MGYG000003060    NP_391417.1  MGYG000003060_00284      NP_249596.1   \n",
       "MGYG000000626    NP_391417.1  MGYG000000626_00410      NP_249596.1   \n",
       "MGYG000002462    NP_391417.1  MGYG000002462_03497      NP_249596.1   \n",
       "MGYG000002532    NP_391417.1  MGYG000002532_03127      NP_249596.1   \n",
       "...                      ...                  ...              ...   \n",
       "MGYG000000864    NP_391417.1  MGYG000000864_02155      NP_249596.1   \n",
       "MGYG000001491    NP_391417.1  MGYG000001491_01283      NP_249596.1   \n",
       "MGYG000001312    NP_391417.1  MGYG000001312_01817      NP_249596.1   \n",
       "MGYG000001486    NP_391417.1  MGYG000001486_01494      NP_249596.1   \n",
       "MGYG000002404    NP_391417.1  MGYG000002404_03669      NP_249596.1   \n",
       "\n",
       "                  fbest_hit_pseudo  \n",
       "MGYG000003619  MGYG000003619_01053  \n",
       "MGYG000003060  MGYG000003060_00284  \n",
       "MGYG000000626  MGYG000000626_00410  \n",
       "MGYG000002462  MGYG000002462_03497  \n",
       "MGYG000002532  MGYG000002532_03127  \n",
       "...                            ...  \n",
       "MGYG000000864  MGYG000000864_02155  \n",
       "MGYG000001491  MGYG000001491_01283  \n",
       "MGYG000001312  MGYG000001312_01817  \n",
       "MGYG000001486  MGYG000001486_01494  \n",
       "MGYG000002404  MGYG000002404_03669  \n",
       "\n",
       "[599 rows x 4 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_subt = pd.merge(best_rHits_subt_df,best_fHits_subt_df, left_index=True, right_index=True)\n",
    "df_pseudo = pd.merge(best_rHits_pseudo_df,best_fHits_pseudo_df, left_index=True, right_index=True)\n",
    "df = pd.merge(df_subt,df_pseudo, left_index=True, right_index=True)\n",
    "bs_csrA_prot_ID = 'NP_391417.1'\n",
    "pa_csrA_prot_ID = 'NP_249596.1'\n",
    "\n",
    "df_filt_s = df.loc[df['rbest_hit_subt'] == bs_csrA_prot_ID,:]\n",
    "df_filt_p = df.loc[df['rbest_hit_pseudo'] == pa_csrA_prot_ID,:]\n",
    "df_double_verified = df_filt_s.loc[df_filt_s['rbest_hit_pseudo'] == pa_csrA_prot_ID,:]\n",
    "df_single_verified = df.loc[set(list(df_filt_s.index) +  list(df_filt_p.index)),:]\n",
    "df_double_verified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a235d5c8-8745-4b24-830c-43640bd929f5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rbest_hit_subt</th>\n",
       "      <th>fbest_hit_subt</th>\n",
       "      <th>rbest_hit_pseudo</th>\n",
       "      <th>fbest_hit_pseudo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MGYG000001571</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000001571_04093</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000001571_04093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000004825</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000004825_00595</td>\n",
       "      <td>NP_251954.1</td>\n",
       "      <td>MGYG000004825_00661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000001275</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000001275_00964</td>\n",
       "      <td>NP_251201.1</td>\n",
       "      <td>MGYG000001275_00045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002416</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000002416_00501</td>\n",
       "      <td>NP_252928.1</td>\n",
       "      <td>MGYG000002416_01080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000003661</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000003661_01348</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000003661_01348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002404</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000002404_03669</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000002404_03669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002344</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000002344_00755</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000002344_00755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002357</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000002357_03771</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000002357_03771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000000620</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000000620_00120</td>\n",
       "      <td>NP_253641.1</td>\n",
       "      <td>MGYG000000620_00564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000001706</th>\n",
       "      <td>NP_391417.1</td>\n",
       "      <td>MGYG000001706_01256</td>\n",
       "      <td>NP_249596.1</td>\n",
       "      <td>MGYG000001706_01256</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>842 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              rbest_hit_subt       fbest_hit_subt rbest_hit_pseudo  \\\n",
       "MGYG000001571    NP_391417.1  MGYG000001571_04093      NP_249596.1   \n",
       "MGYG000004825    NP_391417.1  MGYG000004825_00595      NP_251954.1   \n",
       "MGYG000001275    NP_391417.1  MGYG000001275_00964      NP_251201.1   \n",
       "MGYG000002416    NP_391417.1  MGYG000002416_00501      NP_252928.1   \n",
       "MGYG000003661    NP_391417.1  MGYG000003661_01348      NP_249596.1   \n",
       "...                      ...                  ...              ...   \n",
       "MGYG000002404    NP_391417.1  MGYG000002404_03669      NP_249596.1   \n",
       "MGYG000002344    NP_391417.1  MGYG000002344_00755      NP_249596.1   \n",
       "MGYG000002357    NP_391417.1  MGYG000002357_03771      NP_249596.1   \n",
       "MGYG000000620    NP_391417.1  MGYG000000620_00120      NP_253641.1   \n",
       "MGYG000001706    NP_391417.1  MGYG000001706_01256      NP_249596.1   \n",
       "\n",
       "                  fbest_hit_pseudo  \n",
       "MGYG000001571  MGYG000001571_04093  \n",
       "MGYG000004825  MGYG000004825_00661  \n",
       "MGYG000001275  MGYG000001275_00045  \n",
       "MGYG000002416  MGYG000002416_01080  \n",
       "MGYG000003661  MGYG000003661_01348  \n",
       "...                            ...  \n",
       "MGYG000002404  MGYG000002404_03669  \n",
       "MGYG000002344  MGYG000002344_00755  \n",
       "MGYG000002357  MGYG000002357_03771  \n",
       "MGYG000000620  MGYG000000620_00564  \n",
       "MGYG000001706  MGYG000001706_01256  \n",
       "\n",
       "[842 rows x 4 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_single_verified"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80a31600-e031-4d46-87c8-bb4b1ca8f21e",
   "metadata": {},
   "source": [
    "### Get the protein sequences for the putative csrAs in the Hs gut microbiome species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "03a3cf9f-d86a-4758-80d6-c59bccc83687",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_protein_dict(accessions, all_blast_results_dir):\n",
    "    \"\"\"Return dict with {prot_id:seq}, where seq is the forward blast best hit\"\"\"\n",
    "    seq_dict = {}\n",
    "    query_dir = os.path.join(all_blast_results_dir, 'forward_best_hits')\n",
    "    for acc in accessions:\n",
    "        seq_dict.update(pfa.get_seq_dict(os.path.join(query_dir, acc+'.faa')))\n",
    "    return seq_dict\n",
    "    \n",
    "    \n",
    "seq_dict_pseudo = get_protein_dict(accessions  = list(df_filt_p.index), all_blast_results_dir = '../all_rblast_results_Psuedo')\n",
    "seq_dict_subtilis = get_protein_dict(accessions = list(df_filt_s.index), all_blast_results_dir = '../all_rblast_results_subtilis')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0f19c77-cf69-4649-acce-de8346c51767",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "31f53d2b-903a-4362-8e0d-439afd541df9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def separate_lineage(df):\n",
    "    lins = list(metadata_df['Lineage'])\n",
    "    d = {}\n",
    "    for acc in df.index:\n",
    "        lin = df.loc[acc, 'Lineage']\n",
    "        rank_l =lin.split(';')\n",
    "        d[acc] = []\n",
    "        for rank in rank_l:\n",
    "            d[acc].append(rank[3:]) \n",
    "    lineage_df = pd.DataFrame.from_dict(d, orient='index')\n",
    "    lineage_df.columns = ['Domain', 'Phylum', 'Class', 'Order', 'Fine', 'Gold', 'Species']\n",
    "    return df.merge(lineage_df, left_index=True, right_index=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5e7414fb-76f4-4e2a-82c8-fcc7e09d2ed1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "path_to_csra_tsv = '../Data/csrs.tsv'\n",
    "mot_csrA_df = pd.read_csv(path_to_csra_tsv, sep='\\t')\n",
    "mot_csrA_seqs_d = pfa.get_seq_dict('../Protein_seqs/csrA_sequences.faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e876f090-dc15-4413-b931-6c587ce508b0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_type</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Fine</th>\n",
       "      <th>Gold</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MGYG000002415</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Proteus</td>\n",
       "      <td>Proteus sp001722135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002488</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Proteus</td>\n",
       "      <td>Proteus penneri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000001624</th>\n",
       "      <td>MAG</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Proteus</td>\n",
       "      <td>Proteus cibarius</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002514</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Proteus</td>\n",
       "      <td>Proteus mirabilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000004813</th>\n",
       "      <td>MAG</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Proteus</td>\n",
       "      <td>Proteus sp003144505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002465</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Yersinia</td>\n",
       "      <td>Yersinia massiliensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002335</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Yersinia</td>\n",
       "      <td>Yersinia enterocolitica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000001351</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Edwardsiella</td>\n",
       "      <td>Edwardsiella tarda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002467</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Yersinia</td>\n",
       "      <td>Yersinia frederiksenii_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002524</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Yersinia</td>\n",
       "      <td>Yersinia bercovieri</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Genome_type    Domain          Phylum                Class  \\\n",
       "MGYG000002415     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002488     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000001624         MAG  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002514     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000004813         MAG  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "...                   ...       ...             ...                  ...   \n",
       "MGYG000002465     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002335     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000001351     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002467     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002524     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "\n",
       "                          Order                Fine          Gold  \\\n",
       "MGYG000002415  Enterobacterales  Enterobacteriaceae       Proteus   \n",
       "MGYG000002488  Enterobacterales  Enterobacteriaceae       Proteus   \n",
       "MGYG000001624  Enterobacterales  Enterobacteriaceae       Proteus   \n",
       "MGYG000002514  Enterobacterales  Enterobacteriaceae       Proteus   \n",
       "MGYG000004813  Enterobacterales  Enterobacteriaceae       Proteus   \n",
       "...                         ...                 ...           ...   \n",
       "MGYG000002465  Enterobacterales  Enterobacteriaceae      Yersinia   \n",
       "MGYG000002335  Enterobacterales  Enterobacteriaceae      Yersinia   \n",
       "MGYG000001351  Enterobacterales  Enterobacteriaceae  Edwardsiella   \n",
       "MGYG000002467  Enterobacterales  Enterobacteriaceae      Yersinia   \n",
       "MGYG000002524  Enterobacterales  Enterobacteriaceae      Yersinia   \n",
       "\n",
       "                                Species  \n",
       "MGYG000002415       Proteus sp001722135  \n",
       "MGYG000002488           Proteus penneri  \n",
       "MGYG000001624          Proteus cibarius  \n",
       "MGYG000002514         Proteus mirabilis  \n",
       "MGYG000004813       Proteus sp003144505  \n",
       "...                                 ...  \n",
       "MGYG000002465     Yersinia massiliensis  \n",
       "MGYG000002335   Yersinia enterocolitica  \n",
       "MGYG000001351        Edwardsiella tarda  \n",
       "MGYG000002467  Yersinia frederiksenii_C  \n",
       "MGYG000002524       Yersinia bercovieri  \n",
       "\n",
       "[95 rows x 8 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {}\n",
    "d.update(seq_dict_subtilis)\n",
    "d.update(seq_dict_pseudo)\n",
    "matches = {prot_id:{} for prot_id in mot_csrA_seqs_d}\n",
    "for prot_id, seq in mot_csrA_seqs_d.items():\n",
    "    for gut_prot_id, gut_seq in d.items():\n",
    "        if seq == gut_seq:\n",
    "            matches[prot_id][gut_prot_id] = seq\n",
    "matches_e = {}\n",
    "for mot_csrA_id, dic, in matches.items():\n",
    "    if len(dic)!=0:\n",
    "        for gut_csrA_id, seq in dic.items():\n",
    "            matches_e[gut_csrA_id] = seq\n",
    "species_exact_matches =  [prot_id.split('_')[0] for prot_id in matches_e.keys()]\n",
    "df = metadata_df.loc[species_exact_matches,:]\n",
    "df = separate_lineage(df)[['Genome_type', 'Domain', 'Phylum', 'Class', 'Order', 'Fine', 'Gold', 'Species']]        \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "17cf6935-b3fc-429f-9d3b-c5a17435d103",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Campylobacterota',\n",
       " 'Desulfobacterota',\n",
       " 'Firmicutes',\n",
       " 'Firmicutes_A',\n",
       " 'Proteobacteria'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(list(df['Phylum']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc9afca7-f879-4204-9a4e-0da199759bda",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "154d12d0-bedd-49b5-8932-d9e0a21aaa8f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "61c35e59-2680-455a-85a7-1e3bf60e86cf",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "source": [
    "## Commented out code below should only be run once because of high computational burden"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29ce6997-0cca-41a1-bcb1-4f805009197f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Make one blast database from all Hs gut microbiome representative species  \n",
    "1) Concatenate proteomes from each species into one large file to make the database  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1f5ff714-56cc-4ed1-81b1-dae12e04e2f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# proteomes = os.listdir('../Proteomes/')   \n",
    "# for proteome in proteomes:   \n",
    "#     fp = os.path.join('../Proteomes', proteome)    \n",
    "#     !cat $fp >> '../giant_proteome.faa'  \n",
    "#     !makeblastdb -in ../all_hs_gut_proteomes.faa -dbtype prot -out hs_gut_microbiome_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fe2a0875-8fb0-4b54-8e24-57fa3ad4e158",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #write each mot csrA sequence to its own individual fasta file\n",
    "# p_dir =  '../ind_csrA_fastas'\n",
    "# os.makedirs(p_dir)\n",
    "# for seq_id, seq in mot_csrA_seqs_d.items():\n",
    "#     pfa.write_to_fasta({seq_id:seq}, os.path.join(p_dir, seq_id+'.faa'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd8c5908-bd82-4465-98b3-937b20a945b1",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Blastp each csrA query against the blastp database hs_gut_microbiome_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c16baee0-af48-4498-950e-f20730213fa5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def multiquery_blastp(individ_fasta_dir, out_p_dir,):   \n",
    "    \n",
    "    proteins = os.listdir(individ_fasta_dir)\n",
    "    os.makedirs(p_dir)\n",
    "    for protein in proteins:\n",
    "        fp = os.path.join('../ind_csrA_fastas/', protein)\n",
    "        out_filepath = os.path.join(out_p_dir, protein)\n",
    "        base_filepath = '../hs_gut_microbiome_db/hs_gut_microbiome_db'\n",
    "        print(fp, base_filepath, out_filepath)\n",
    "        !blastp -query $fp -db $base_filepath -num_descriptions 50 -num_alignments 50 -word_size 7 -out $out_filepath\n",
    "        print(f'finished blasting {protein} ')\n",
    "        \n",
    "        \n",
    "out_p_dir = '../All_csrA_hits'\n",
    "individ_fasta_dir = '../ind_csrA_fastas/'\n",
    "#multiquery_blastp(individ_fasta_dir, out_p_dir,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1cf49696-fda6-44bc-bed3-4b081c2a78c0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_perc_id_blast(fp, cutoff = 99):\n",
    "    \"\"\"Return the top hit and the percent identity to the query\"\"\"\n",
    "    with open(fp, 'r') as f: \n",
    "        a = f.readlines()\n",
    "    d = {}\n",
    "    \n",
    "    for i, ele in enumerate(a):\n",
    "        if ele.startswith('>'):\n",
    "            prot_id = a[i].split(' ')[0][1:]\n",
    "            perc_ident = float(a[i+4].split(',')[0].split(' ')[-1].replace('(','').replace(')','')[:-1])\n",
    "            if perc_ident>=cutoff:\n",
    "                d[prot_id] = perc_ident\n",
    "\n",
    "    return d\n",
    "\n",
    "\n",
    "per_id_d = {}\n",
    "p_dir = '../All_csrA_hits/'\n",
    "files = os.listdir(p_dir)\n",
    "for file in files:\n",
    "    fp = os.path.join(p_dir, file)\n",
    "    per_id_d[file.replace('.faa', '')] = get_perc_id_blast(fp, cutoff = 100)\n",
    "df_blast_species = pd.DataFrame.from_dict(per_id_d, orient='index').transpose()\n",
    "blast_species = set([ele.split('_')[0] for ele in list(df_blast_species.index)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "85446a7e-fc77-4728-9dc3-d19ea901140e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_type</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Fine</th>\n",
       "      <th>Gold</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MGYG000002458</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Yersinia</td>\n",
       "      <td>Yersinia aleksiciae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002519</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Enterobacter</td>\n",
       "      <td>Enterobacter asburiae_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002335</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Yersinia</td>\n",
       "      <td>Yersinia enterocolitica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000004813</th>\n",
       "      <td>MAG</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Proteus</td>\n",
       "      <td>Proteus sp003144505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002507</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Escherichia albertii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002520</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Pluralibacter</td>\n",
       "      <td>Pluralibacter gergoviae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000001624</th>\n",
       "      <td>MAG</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Proteus</td>\n",
       "      <td>Proteus cibarius</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002816</th>\n",
       "      <td>MAG</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Citrobacter</td>\n",
       "      <td>Citrobacter braakii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000000110</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Kluyvera</td>\n",
       "      <td>Kluyvera sp902363335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGYG000002357</th>\n",
       "      <td>Isolate</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillus licheniformis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Genome_type    Domain          Phylum                Class  \\\n",
       "MGYG000002458     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002519     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002335     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000004813         MAG  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002507     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "...                   ...       ...             ...                  ...   \n",
       "MGYG000002520     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000001624         MAG  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002816         MAG  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000000110     Isolate  Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "MGYG000002357     Isolate  Bacteria      Firmicutes              Bacilli   \n",
       "\n",
       "                          Order                Fine           Gold  \\\n",
       "MGYG000002458  Enterobacterales  Enterobacteriaceae       Yersinia   \n",
       "MGYG000002519  Enterobacterales  Enterobacteriaceae   Enterobacter   \n",
       "MGYG000002335  Enterobacterales  Enterobacteriaceae       Yersinia   \n",
       "MGYG000004813  Enterobacterales  Enterobacteriaceae        Proteus   \n",
       "MGYG000002507  Enterobacterales  Enterobacteriaceae    Escherichia   \n",
       "...                         ...                 ...            ...   \n",
       "MGYG000002520  Enterobacterales  Enterobacteriaceae  Pluralibacter   \n",
       "MGYG000001624  Enterobacterales  Enterobacteriaceae        Proteus   \n",
       "MGYG000002816  Enterobacterales  Enterobacteriaceae    Citrobacter   \n",
       "MGYG000000110  Enterobacterales  Enterobacteriaceae       Kluyvera   \n",
       "MGYG000002357        Bacillales         Bacillaceae       Bacillus   \n",
       "\n",
       "                               Species  \n",
       "MGYG000002458      Yersinia aleksiciae  \n",
       "MGYG000002519  Enterobacter asburiae_B  \n",
       "MGYG000002335  Yersinia enterocolitica  \n",
       "MGYG000004813      Proteus sp003144505  \n",
       "MGYG000002507     Escherichia albertii  \n",
       "...                                ...  \n",
       "MGYG000002520  Pluralibacter gergoviae  \n",
       "MGYG000001624         Proteus cibarius  \n",
       "MGYG000002816      Citrobacter braakii  \n",
       "MGYG000000110     Kluyvera sp902363335  \n",
       "MGYG000002357   Bacillus licheniformis  \n",
       "\n",
       "[103 rows x 8 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_species_df = separate_lineage(metadata_df.loc[blast_species,:])[['Genome_type', 'Domain', 'Phylum', 'Class', 'Order', 'Fine', 'Gold', 'Species']]\n",
    "blast_species_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b17f98c-638b-4227-9085-06db98949787",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6714c79c-abad-4a58-8c79-a7e4c23c4005",
   "metadata": {},
   "source": [
    "The dominant gut microbial phyla are   \n",
    "Firmicutes, Bacteroidetes, Actinobacteria, Proteobacteria, Fusobacteria, and Verrucomicrobia,  \n",
    "with the two phyla Firmicutes and Bacteroidetes [13] representing 90% of gut microbiota.  \n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6351938/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0526425b-83a6-498d-b100-b0c15fd3fb59",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'',\n",
       " 'Acetivibrio_A ethanolgignens',\n",
       " 'Agathobacter rectalis',\n",
       " 'Agathobacter sp000434275',\n",
       " 'Agathobacter sp900546625',\n",
       " 'Agathobacter sp900550545',\n",
       " 'Agathobaculum sp900291975',\n",
       " 'Anaerotruncus colihominis',\n",
       " 'Bacillus licheniformis',\n",
       " 'Bacillus paralicheniformis',\n",
       " 'CAG-303 sp000437755',\n",
       " 'CAG-411 sp000437275',\n",
       " 'Campylobacter_D jejuni',\n",
       " 'Citrobacter braakii',\n",
       " 'Citrobacter europaeus',\n",
       " 'Citrobacter freundii',\n",
       " 'Citrobacter gillenii',\n",
       " 'Citrobacter portucalensis',\n",
       " 'Citrobacter portucalensis_A',\n",
       " 'Citrobacter youngae',\n",
       " 'Citrobacter_A amalonaticus',\n",
       " 'Citrobacter_A farmeri',\n",
       " 'Citrobacter_A sedlakii',\n",
       " 'Citrobacter_B koseri',\n",
       " 'Cronobacter malonaticus',\n",
       " 'Cronobacter sakazakii',\n",
       " 'Edwardsiella tarda',\n",
       " 'Eisenbergiella tayi',\n",
       " 'Enterobacter asburiae_B',\n",
       " 'Enterobacter cancerogenus',\n",
       " 'Enterobacter chengduensis',\n",
       " 'Enterobacter cloacae',\n",
       " 'Enterobacter cloacae_K',\n",
       " 'Enterobacter hormaechei_A',\n",
       " 'Enterobacter kobei',\n",
       " 'Enterobacter ludwigii',\n",
       " 'Enterobacter mori',\n",
       " 'Enterobacter roggenkampii',\n",
       " 'Enterobacter sichuanensis',\n",
       " 'Enterobacter_A timonensis',\n",
       " 'Enterobacter_D kobei_A',\n",
       " 'Escherichia albertii',\n",
       " 'Escherichia coli_D',\n",
       " 'Escherichia fergusonii',\n",
       " 'Escherichia marmotae',\n",
       " 'Escherichia sp000208585',\n",
       " 'Eubacterium_F sp000434115',\n",
       " 'Exiguobacterium sp902362975',\n",
       " 'Flavonifractor plautii',\n",
       " 'Helicobacter pylori_C',\n",
       " 'Helicobacter_D pullorum',\n",
       " 'Klebsiella aerogenes',\n",
       " 'Klebsiella pneumoniae',\n",
       " 'Klebsiella quasipneumoniae',\n",
       " 'Klebsiella variicola',\n",
       " 'Klebsiella_A grimontii',\n",
       " 'Klebsiella_A indica',\n",
       " 'Klebsiella_A michiganensis',\n",
       " 'Klebsiella_A oxytoca',\n",
       " 'Kluyvera ascorbata_B',\n",
       " 'Kluyvera sp902363335',\n",
       " 'Lachnospira sp900316325',\n",
       " 'Nitratidesulfovibrio vulgaris',\n",
       " 'Paenibacillus lautus_A',\n",
       " 'Phytobacter massiliensis',\n",
       " 'Phytobacter sp002377245',\n",
       " 'Pluralibacter gergoviae',\n",
       " 'Proteus cibarius',\n",
       " 'Proteus mirabilis',\n",
       " 'Proteus penneri',\n",
       " 'Proteus sp001722135',\n",
       " 'Proteus sp003144505',\n",
       " 'Pseudescherichia sp002298805',\n",
       " 'Pseudescherichia vulneris',\n",
       " 'Pseudomonas aeruginosa',\n",
       " 'Pseudomonas citronellolis',\n",
       " 'Pseudomonas_E extremaustralis',\n",
       " 'Raoultella ornithinolytica',\n",
       " 'Roseburia sp900542495',\n",
       " 'Roseburia sp900552665',\n",
       " 'Ruminiclostridium_E siraeum',\n",
       " 'Ruminiclostridium_E sp003512525',\n",
       " 'Ruminiclostridium_E sp900539195',\n",
       " 'Salmonella bongori',\n",
       " 'Salmonella enterica',\n",
       " 'Succinatimonas hippei',\n",
       " 'Superficieibacter sp900766525',\n",
       " 'TF01-11 sp003524945',\n",
       " 'Vibrio cholerae',\n",
       " 'Vibrio metoecus',\n",
       " 'Virgibacillus kapii',\n",
       " 'Yersinia aleksiciae',\n",
       " 'Yersinia bercovieri',\n",
       " 'Yersinia enterocolitica',\n",
       " 'Yersinia frederiksenii',\n",
       " 'Yersinia frederiksenii_C',\n",
       " 'Yersinia intermedia',\n",
       " 'Yersinia kristensenii',\n",
       " 'Yersinia kristensenii_B',\n",
       " 'Yersinia massiliensis',\n",
       " 'Yersinia mollaretii',\n",
       " 'Yersinia pestis',\n",
       " 'Yokenella regensburgei'}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(blast_species_df['Species'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "717945f6-c506-4ea0-919c-a100fe2930b4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d2c88fd-6422-49ab-a3cf-be759ee9af2e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "57432706-ede6-4094-9b56-00f6038aa5d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "305c051e-9944-490d-b463-02b254be0f9f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f06b60b6-62bd-414d-a39c-552a8da3e86e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85332f54-c8ea-47cb-a17b-b2151e508b33",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cc4e5a8-7c83-42f3-80c5-046cfefb31d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23a43136-e8b8-4a3c-8148-7753244acf4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c27938cb-b1fd-40d6-a351-9f8f982e92fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38d67047-3a06-4215-a99f-121f3d879530",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4e859c4-c056-48c2-8d4b-c00b722cd2ac",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
